pkg/volume/util/util.go

```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
+		// for multi replica volumes, need zones and not zone
+		return nil, fmt.Errorf("for multi replica volumes, single Zone cannot be specified; require Zones")
+	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DynamicProvisioningScheduling) {
+		// not much to do - return the passed parameters as is
+		return &zoneCfg, nil
+	}
+
+	allowedZones, err := ZonesFromAllowedTopologies(allowedTopologies)
+	if err != nil {
+		return nil, err
+	}
+	if allowedZones.Len() > 0 {
+		if zonePresent || zonesPresent {
+			return nil, fmt.Errorf("Zone/Zones cannot be specified in StorageClass if allowedTopologies specified")
+		}
+		if allowedZones.Len() < replicas {
+			// for multi replica volumes, need as many allowedZones as replicas
+			return nil, fmt.Errorf("Require at least %d allowedZones for volumes with %d replicas", replicas, replicas)
+		}
+	}
+
+	if node == nil {
+		// node not specified, so derive zone/zones based on allowedZones
+		if allowedZones.Len() == 1 {
+			zoneCfg.ZonePresent = true
+			zoneCfg.Zone = allowedZones.UnsortedList()[0]
+		}
+		if allowedZones.Len() > 1 {
+			zoneCfg.ZonesPresent = true
+			zoneCfg.Zones = strings.Join(allowedZones.List(), ",")
+		}
+		return &zoneCfg, nil
+	}
+
+	// non-null node is specified from this point on
+	if zonePresent || zonesPresent {
+		return nil, fmt.Errorf("Zone/Zones cannot be specified in StorageClass if VolumeBindingMode is set to WaitForFirstConsumer. Please specify AllowedTopology in StorageClass for constraining zones.")
+	}
+
+	// we pick node's zone and ignore any allowedTopologies (since scheduler has accounted for it already)
+	z, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
+	if !ok {
+		return nil, fmt.Errorf("%s Label for node missing", kubeletapis.LabelZoneFailureDomain)
+	}
+	zoneCfg.ReplicaZoneFromNode = z
+	zoneCfg.ReplicaZoneFromNodePresent = true
+
+	if replicas < 1 {
+		return nil, fmt.Errorf("Invalid number of replicas passed for Volume: %d", replicas)
+	} else if replicas == 1 {
+		// consumers can use either Zone or ReplicaZoneFromNode since both will be the same with single replicas
+		zoneCfg.ZonePresent = true
+		zoneCfg.Zone = zoneCfg.ReplicaZoneFromNode
+	} else {
+		// more than one replicas i.e. multi zonal volumes
+		if allowedZones.Len() > 1 {
+			// erase ReplicaZoneFromNode from zones since consumers will pick an element from Zones and it should not be ReplicaZoneFromNode
+			allowedZones.Delete(zoneCfg.ReplicaZoneFromNode)
+			zoneCfg.ZonesPresent = true
+			zoneCfg.Zones = strings.Join(allowedZones.List(), ",")
+		}
+	}
+	return &zoneCfg, nil
+}
+
+// ZonesFromAllowedTopologies returns a list of zones specified in allowedTopologies
+func ZonesFromAllowedTopologies(allowedTopologies []v1.TopologySelectorTerm) (*sets.String, error) {
+	zones := make(sets.String)
+	for _, term := range allowedTopologies {
+		for _, exp := range term.MatchLabelExpressions {
+			if exp.Key == kubeletapis.LabelZoneFailureDomain {
```

Comment:
We should document somewhere that only `LabelZoneFailureDomain` is supported on AWS (in a separate PR).